poa-token/src/contract.rs

```diff
@@ -56,6 +153,236 @@ impl Contract {
+            ),
+        };
+
+        let self_total_supply = self.ft_total_supply();
+        require!(
+            parsed_balance >= self_total_supply,
+            format!(
+                "Migration required that the wrapped token have sufficient balance to cover for the balance in this contract: {} < {}",
+                parsed_balance.0, self_total_supply.0
+            )
+        );
+
+        *self.wrapped_token_mut() = Some(token_account_id);
+    }
+
+    /// Returns the amount of tokens that were used/unwrapped after requesting an unwrap
+    #[private]
+    pub fn ft_resolve_unwrap(
+        &mut self,
+        sender_id: &AccountId,
+        amount: u128,
+        is_call: bool,
+    ) -> U128 {
+        let used = match env::promise_result(0) {
+            near_sdk::PromiseResult::Successful(value) => {
+                if let Ok(unused_amount) = near_sdk::serde_json::from_slice::<U128>(&value) {
+                    std::cmp::min(amount, unused_amount.0)
+                } else {
+                    amount
+                }
+            }
+            near_sdk::PromiseResult::Failed => {
+                if is_call {
+                    // do not refund on failed `ft_transfer_call` due to
+                    // NEP-141 vulnerability: `ft_resolve_transfer` fails to
+                    // read result of `ft_on_transfer` due to insufficient gas
+                    amount
+                } else {
+                    0
+                }
+            }
+        };
+
+        let to_refund = amount.saturating_sub(used);
+        self.token_mut().internal_deposit(sender_id, to_refund);
+
+        used.into()
+    }
+}
+
+impl Contract {
+    fn unwrap_and_transfer(
+        &mut self,
+        token_destination: AccountId,
+        amount: U128,
+        memo: Option<String>,
+        msg: Option<String>,
+    ) -> PromiseOrValue<U128> {
+        let Some(wrapped_token_id) = self.wrapped_token().cloned() else {
+            env::panic_str("Unwrapping is only for wrapped tokens");
+        };
+
+        let previous_owner = &*PREDECESSOR_ACCOUNT_ID;
+
+        self.ft_withdraw(previous_owner, amount, None);
+
+        // There seems to be a bug in clippy... this isn't a single match arm
+        #[allow(clippy::single_match_else)]
+        match msg {
+            Some(inner_msg) => ext_ft_core::ext(wrapped_token_id.clone())
+                .with_attached_deposit(NearToken::from_yoctonear(1))
+                .ft_transfer_call(token_destination, amount, memo, inner_msg)
+                .then(
+                    Contract::ext(CURRENT_ACCOUNT_ID.clone())
+                        .with_static_gas(FT_UNWRAP_GAS)
+                        .ft_resolve_unwrap(previous_owner, amount.0, true),
+                )
+                .into(),
+            None => ext_ft_core::ext(wrapped_token_id.clone())
+                .with_attached_deposit(NearToken::from_yoctonear(1))
+                .ft_transfer(token_destination, amount, memo)
```

Comment:
Seems like you forget the callback `ft_resolve_unwrap(sender_id, amount, false)`